File: 0144.binary_tree_preorder_traversal/solution.py
from typing import List

import utils


class Solution:
    def preorder_traversal(self, root: utils.TreeNode) -> List[int]:
        stack, ans = [root], []
        while stack:
            node = stack.pop()
            if node:
                ans.append(node.val)
                if node.right:
                    stack.append(node.right)
                if node.left:
                    stack.append(node.left)
        return ans

File: 0144.binary_tree_preorder_traversal/solution.go
package leetcode0144

import "github.com/Ackerr/Algorithms/utils"

func preorderTraversal(root *utils.TreeNode) []int {
	stack := []*utils.TreeNode{root}
	var ans []int
	for len(stack) > 0 {
		index := len(stack) - 1
		node := stack[index]
		stack = stack[:index]
		if node != nil {
			ans = append(ans, node.Val)
			if node.Right != nil {
				stack = append(stack, node.Right)
			}
			if node.Left != nil {
				stack = append(stack, node.Left)
			}
		}
	}
	return ans
}
